src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -2515,6 +2515,11 @@ public static NodeMigrationData MigrateShortestLacingToAutoLacing(NodeMigrationD
 
         #endregion
 
+        public virtual void OnBeforeAdd(Dynamo.Models.DynamoModel dynamoModel, WorkspaceModel workspace)
```

Comment:
this is quite a big change - in the rest of Dynamo it's common for models to only have references to their children - ie DynamoModel knows about NodeModels - and NodeModels know about portModels, but not the other way around.

Curious why you need this reference - maybe it will become clear.

Reply:
hmm - it makes a lot of sense that the python nodeModel needs to continue working even in headless Dynamo.

Reply:
I'm curious what you think of using the active `ExecutionSession` instead - storing the pythonEngine there and extracting it from the static map when the node is constructed.

Reply:
I'm open to change this. `ExecutingSession` is looking uglier each time we add another new thing to it, but I guess it can be used.

This does expose more data, but we are not being super consistent with the visibility anyway. For view models the rules are pretty different than what you mention here.

I don't think this would break headless Dynamo, unless you added a node to a graph dynamically (can you do that headless?) and forgot to assign an `Engine` to it.

Reply:
you can add a node to a graph dynamically in headless mode for sure - consider some headless client of Dynamo that calls the createNodeCommand - but - I was actually trying to make the point that I thought your solution was better than my original suggestion because this doesn't require the NodeView to be around to correctly set the engine.

Reply:
for some context - I believe historically, there was a giant static class called something like `DynamoSettings` or `DynamoModelEntryPoint` that was used from everywhere, this made the separation between core and wpf layers very difficult to enforce, and a big refactor was undertaken.

There was also the idea that one day we'd have multiple DynamoModels running at the same time in the same process, but this hit other roadblocks.

I'm not sure the pain for nodeModel developers (like you here) is worth the old ideas from that refactor any longer.

Reply:
Multiple DynamoModel sounds intriguing. In any case a node would belong to just one of those, right? 

Giving a node model access to its parent workspace model and that one having access to its parent dynamo model does not sound so bad to me. It feels having access to the context a node model is in should be more direct than relying on global classes. 

In any case, we probably don't want to make a refactor part of this PR. I'll stick to `ExecutingSession` for now.

Reply:
Now thinking again from a security perspective, accessing the `DynamoModel` opens the door for a node to do a bunch of things we may not want them to do. Note that the risk currently exists for a custom `NodeView` as they can access the `DynamoViewModel` and then to the `DynamoModel`.

It might be a  good idea to think this further and consider refactoring in 3.0.